### PR TITLE
fix(`forge`): Dedup remappings on build correctly & do not set context by default

### LINKED ISSUE #5447: bug(forge): overlapping nested remappings are not resolved
### Component

Forge

### Have you ensured that all of these are up to date?

- [x] Foundry
- [x] Foundryup

### What version of Foundry are you on?

`forge 0.2.0 (02e430c 2023-07-21T00:24:08.481175000Z)`

### What command(s) is the bug in?

`forge build`

### Operating System

macOS Silicon

## Problem

#5397 added support for remapping contexts (closing #1855), however, the implementation does not handle all edge cases yet.

Suppose you have a project structure like this:

```tree
project/
└── lib/
    ├── foo/
    │   └── lib/
    │       └── openzeppelin-contracts  
    └── bar/
        └── lib/
            └── openzeppelin-contracts
```

In both `foo` and `bar`, `openzeppelin-contracts` is remapped to the following:

```bash
@openzeppelin/contracts/=lib/openzeppelin-contracts/contracts/
```

The bug is that only one of the two remappings gets resolved, not both.

## Error Log

Sneak peek:

>  Unable to resolve imports: "@openzeppelin/contracts/token/ERC20/IERC20.sol"

See the full log [here](https://app.warp.dev/block/NzkoqBiRAVrpY7IW8d5M0r).

The issue is that only the `v2-core` remapping appears in the auto-detected remappings:

```bash
lib/v2-core:@openzeppelin/contracts/=/Users/prb/Sablier/examples/lib/v2-core/lib/openzeppelin-contracts/contracts/
```

## Reproduce

1. Clone [sablier-labs/examples](https://github.com/sablier-labs/examples) at commit [eae781](https://github.com/sablier-labs/examples/commit/eae781627e15d963a62c1545fe0944ad94d5519c)
2. Remove the `permit2` remapping in `remappings.txt`
3. Run `forge build`

### LINKED ISSUE #5475: bug: regression introduced in PR#5397 (remapping contexts)
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [ ] Foundryup

### What version of Foundry are you on?

_No response_

### What command(s) is the bug in?

forge build

### Operating System

None

### Describe the bug

Seems https://github.com/foundry-rs/foundry/pull/5397 was inadvertently a breaking change. Steps to reproduce:
1. `git clone https://github.com/gitcoinco/Alpha-Governor-Upgrade`
2. `git checkout 17f7717`
3. `forge build` the repo using foundry commit `d231a90f`: build succeeds (this is the commit right before that PR)
3. `forge build` the repo using foundry commit `8e365bee`: build fails (this is the commit where that PR was merged)

### PR DESCRIPTION
**Pull request by Evalir**
## Motivation

Right now we have an issue where remappings are being set even if they're optional—which blows up the whole input, and creates problems for _nested_ contexts, which we yet can't really handle. There's also bad deduping happening as we're not taking contexts into account.

## Solution

Do not set contexts by default which resolves #5475.

Also dedupe and sort remappings correctly, which resolves #5447.

### CODE REVIEW

File: cli/src/cmd/forge/build/core.rs [hunk 1]
```diff
@@ -153,7 +153,7 @@ impl<'a> From<&'a CoreBuildArgs> for Figment {
         remappings
             .extend(figment.extract_inner::<Vec<Remapping>>("remappings").unwrap_or_default());
         remappings.sort_by(|a, b| a.name.cmp(&b.name));
-        remappings.dedup_by(|a, b| a.name.eq(&b.name));
+        remappings.dedup_by(|a, b| (&a.context, &a.name).eq(&(&b.context, &b.name)));
```
Comment (mattsse) [inline cli/src/cmd/forge/build/core.rs:156]: this looks correct we don't want duplicate remappings with same context and name.

would be great to have a test for this

Comment (Evalir) [reply cli/src/cmd/forge/build/core.rs:156]: Just to double check here: We don't wanna duplicate remappings that have _both_ the same context and name. Them having same name but different contexts should be OK

File: cli/src/cmd/forge/build/core.rs [hunk 2]
```diff
@@ -153,7 +153,7 @@ impl<'a> From<&'a CoreBuildArgs> for Figment {
         remappings
             .extend(figment.extract_inner::<Vec<Remapping>>("remappings").unwrap_or_default());
         remappings.sort_by(|a, b| a.name.cmp(&b.name));
```
Comment (mattsse) [inline cli/src/cmd/forge/build/core.rs:155]: we also need to sort exactly like:

https://github.com/foundry-rs/foundry/blob/1058c8afb2f5b682d4f40b836c55006734e06945/config/src/providers/remappings.rs#L129-L130

**Review by mattsse — CHANGES_REQUESTED** [review]

File: config/src/providers/remappings.rs [hunk 3]
```diff
@@ -165,16 +165,8 @@ impl<'a> RemappingsProvider<'a> {
...
-                    .into_iter()
-                    .map(Remapping::from)
-                    .map(|mut r| {
-                        // todo: windows/mac/linux
-                        r.context = Some(lib.to_string_lossy().to_string());
-                        r
-                    })
-                    .collect::<Vec<Remapping>>();
+                let mut remappings =
+                    config.remappings.into_iter().map(Remapping::from).collect::<Vec<Remapping>>();
```
Comment (mattsse) [inline config/src/providers/remappings.rs:169]: can we leave a doc here that we eventually could preset the context at this location?

**Pull request merged**